pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go

```diff
@@ -150,6 +153,106 @@ func TestFindAndAddNewPods_WithRescontructedVolume(t *testing.T) {
 
 }
 
+func TestFindAndAddNewPods_WithDifferentConditions(t *testing.T) {
+	tests := []struct {
+		desc           string
+		hasAddedPods   bool
+		expectedFound  bool // Found pod is added to DSW
+		setRemoved     bool
+		setTerminating bool
+	}{
+		{
+			"HasAddedPods is false, ShouldPodRuntimeBeRemoved is true",
+			false,
+			true, // Pod should be added to DSW
+			true,
+			false,
+		},
+		{
+			"HasAddedPods is true, ShouldPodContainersBeTerminating is true",
+			true,
+			true, // Pod should be added to DSW
+			false,
+			true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Set up
+			dswp, fakePodState, pod, expectedVolumeName, pv := prepareDSWPWithPodPV(t)
+			podName := util.GetUniquePodName(pod)
+
+			//let the pod be terminated
+			podGet, exist := dswp.podManager.GetPodByName(pod.Namespace, pod.Name)
+			if !exist {
+				t.Fatalf("Failed to get pod by pod name: %s and namespace: %s", pod.Name, pod.Namespace)
+			}
+			podGet.Status.Phase = v1.PodFailed
+			dswp.podManager.DeletePod(pod)
+			fakePodState.removed = map[kubetypes.UID]struct{}{pod.UID: {}}
+
+			// Add the volume to ASW by reconciling.
+			fakeASW := dswp.actualStateOfWorld
+			reconcileASW(fakeASW, dswp.desiredStateOfWorld, t)
+
+			// Mark the volume as uncertain
+			opts := operationexecutor.MarkVolumeOpts{
```

Comment:
Is this required? I'm not sure if uncertain volume handling logic factors in here

Reply:
I was trying to do the same set up as the reconstruction case below but it did not work. Just changed the set up again.